saml/src/main/java/com/linecorp/armeria/server/saml/HttpRedirectBindingUtil.java

```diff
@@ -0,0 +1,341 @@
+            inflaterOutputStream.write(base64decoded);
+        } catch (IOException e) {
+            throw new SamlException("failed to inflate a SAML message", e);
+        }
+
+        return SamlMessageUtil.unmarshalMessage(bytesOut.toByteArray());
+    }
+
+    /**
+     * Converts an {@link AggregatedHttpMessage} which is received from the remote entity to
+     * a {@link SAMLObject}.
+     */
+    @SuppressWarnings("unchecked")
+    static <T extends SAMLObject> MessageContext<T> toSamlObject(
+            AggregatedHttpMessage msg, String name,
+            Map<String, SamlIdentityProviderConfig> idpConfigs,
+            @Nullable SamlIdentityProviderConfig defaultIdpConfig) throws SamlException {
+        requireNonNull(msg, "msg");
+        requireNonNull(name, "name");
+        requireNonNull(idpConfigs, "idpConfigs");
+
+        final SamlParameters parameters = new SamlParameters(msg);
+        final T message = (T) fromDeflatedBase64(parameters.getFirstValue(name));
+
+        final MessageContext<T> messageContext = new MessageContext<>();
+        messageContext.setMessage(message);
+
+        final Issuer issuer;
+        if (message instanceof RequestAbstractType) {
+            issuer = ((RequestAbstractType) message).getIssuer();
+        } else if (message instanceof StatusResponseType) {
+            issuer = ((StatusResponseType) message).getIssuer();
+        } else {
+            throw new SamlException("invalid message type: " + message.getClass().getSimpleName());
+        }
+
+        // Use the default identity provider config if there's no issuer.
+        final SamlIdentityProviderConfig config;
+        if (issuer != null) {
+            final String idpEntityId = issuer.getValue();
+            config = idpConfigs.get(idpEntityId);
+            if (config == null) {
+                throw new SamlException("a message from unknown identity provider: " + idpEntityId);
+            }
+        } else {
+            if (defaultIdpConfig == null) {
+                throw new SamlException("failed to get an Issuer element");
+            }
+            config = defaultIdpConfig;
+        }
+
+        // If this message is sent via HTTP-redirect binding protocol, its signature parameter should
+        // be validated.
+        validateSignature(config.signingCredential(), parameters, name);
+
+        final String relayState = parameters.getFirstValueOrNull(RELAY_STATE);
+        if (relayState != null) {
+            final SAMLBindingContext context = messageContext.getSubcontext(SAMLBindingContext.class, true);
+            assert context != null;
+            context.setRelayState(relayState);
+        }
+
+        return messageContext;
+    }
+
+    /**
+     * Build a query string which does not start with '?'.
+     */
+    static final class QueryStringBuilder {
+        private static final String UTF_8 = StandardCharsets.UTF_8.name();
```

Comment:
Perhaps we could use Netty's `QueryStringEncoder` and append a substring? e.g.

```java
QueryStringEncoder qse = new QueryStringEncoder("");
qse.addParam();
qse.toString().substring(1);
...

```

Reply:
Fixed. Thanks!